pkg/controllers/dnspolicy/dnspolicy_dnsrecords.go

```diff
@@ -137,3 +135,81 @@ func (r *DNSPolicyReconciler) deleteGatewayDNSRecords(ctx context.Context, gatew
 	}
 	return nil
 }
+
+func (r *DNSPolicyReconciler) buildClusterGateway(ctx context.Context, downstreamClusterName string, clusterAddress []gatewayv1beta1.GatewayAddress) (dns.ClusterGateway, error) {
+	var target dns.ClusterGateway
+	singleClusterAddresses := make([]gatewayv1beta1.GatewayAddress, len(clusterAddress))
+
+	mc := &clusterv1.ManagedCluster{}
+	if err := r.Client().Get(ctx, client.ObjectKey{Name: downstreamClusterName}, mc, &client.GetOptions{}); err != nil {
+		return target, err
+	}
+
+	for i, addr := range clusterAddress {
+		addrType := gatewayv1beta1.IPAddressType
+		if *addr.Type == gateway.MultiClusterHostnameAddressType {
+			addrType = gatewayv1beta1.HostnameAddressType
+		}
+
+		singleClusterAddresses[i] = gatewayv1beta1.GatewayAddress{
+			Type:  &addrType,
+			Value: addr.Value,
+		}
+	}
+
+	target = *dns.NewClusterGateway(mc, singleClusterAddresses)
+
+	return target, nil
+}
+
+func getClusterGatewayAddresses(gw *gatewayv1beta1.Gateway) map[string][]gatewayv1beta1.GatewayAddress {
+	clusterAddrs := make(map[string][]gatewayv1beta1.GatewayAddress, len(gw.Status.Addresses))
+
+	for _, address := range gw.Status.Addresses {
+		var gatewayAddresses []gatewayv1beta1.GatewayAddress
+
+		//Default to Single Cluster (Normal Gateway Status)
+		cluster := "none"
+		addressValue := address.Value
+
+		//Check for Multi Cluster (MGC Gateway Status)
+		if *address.Type == gateway.MultiClusterIPAddressType || *address.Type == gateway.MultiClusterHostnameAddressType {
+			tmpCluster, tmpAddress, found := strings.Cut(address.Value, "/")
+			//If this fails something is wrong and the value hasn't been set correctly
+			if found {
+				cluster = tmpCluster
+				addressValue = tmpAddress
+			}
+		}
+
+		gatewayAddresses = append(gatewayAddresses, gatewayv1beta1.GatewayAddress{
+			Type:  address.Type,
+			Value: addressValue,
+		})
+		clusterAddrs[cluster] = gatewayAddresses
+	}
+
+	return clusterAddrs
+}
+
+func listenerTotalAttachedRoutes(upstreamGateway *gatewayv1beta1.Gateway, downstreamCluster string, specListener gatewayv1beta1.Listener, addresses []gatewayv1beta1.GatewayAddress) int {
+	for _, statusListener := range upstreamGateway.Status.Listeners {
+		// assuming all adresses of the same type on the gateway
+		// for Multi Cluster (MGC Gateway Status)
+		if *addresses[0].Type == gateway.MultiClusterIPAddressType || *addresses[0].Type == gateway.MultiClusterHostnameAddressType {
+			clusterName, listenerName, found := strings.Cut(string(statusListener.Name), ".")
+			if !found {
+				return 0
+			}
+			if clusterName == downstreamCluster && (listenerName == string(specListener.Name) || strings.Contains(string(*specListener.Hostname), "*")) {
```

Comment:
I'm not seeing why we need to check the host for it being a wildcard, you just need to know if the listener has attached routes or not. Are you trying to get round an issue you are seeing, or was this based on some other bit of code that was being used previously?

Reply:
in here I'm matching listeners from spec to the status block to see if I have routes attached. The match is based on the listener's name and the cluster - this is how I find the correct one. In the case of the wildcard, I want to match all of the listeners on the cluster. It is based on the assumption that the wildcard is of a "top level" that could match any listener, I'm returning routes of the first listener.

Reply:
or you are saying that in case of the wildcard we will have the spec name of the `wildcard` that would correspond with the same name in the status of the gateway? i.e. `workload-cluster-1.wildcard`?

Reply:
Why do you think you need to do anything special if it has a wildcard hostname? You just need to find the listener by name and return the attached routes value for it. The gateway controller updating the status will deal with setting the values, we don't need to put our own logic around it.

Reply:
changed. And updated tests to explicitly reflect desired behaviour